internal/pkg/service/buffer/stats/manager.go

```diff
@@ -0,0 +1,104 @@
+package stats
+
+import (
+	"context"
+	"time"
+
+	"github.com/benbjohnson/clock"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/key"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/model"
+)
+
+type Manager struct {
+	ch       chan update
+	syncFn   syncFn
+	clock    clock.Clock
+	ticker   *clock.Ticker
+	perSlice map[key.SliceStatsKey]*stats
+}
+
+type syncFn func(context.Context, []model.SliceStats)
+
+type update struct {
+	key            key.SliceStatsKey
+	size           uint64
+	lastReceivedAt time.Time
+}
+
+type stats struct {
+	count          uint64
+	size           uint64
+	lastReceivedAt time.Time
+	changed        bool
+}
+
+func New(store *store.Store) Manager {
+	clock := clock.New()
+	return Manager{
+		// channel needs to be large enough to not block under average load
+		ch:       make(chan update, 2048),
+		syncFn:   syncToStore(store),
+		clock:    clock,
+		ticker:   clock.Ticker(time.Second),
+		perSlice: make(map[key.SliceStatsKey]*stats),
+	}
+}
+
+func (m *Manager) Notify(key key.SliceStatsKey, size uint64) {
+	m.ch <- update{
+		key:            key,
+		size:           size,
+		lastReceivedAt: m.clock.Now(),
+	}
+}
+
+func (m *Manager) Watch(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				m.ticker.Stop()
+				return
+			case update := <-m.ch:
+				m.handleUpdate(update)
+			case <-m.ticker.C:
+				m.handleSync(ctx)
+			}
+		}
+	}()
+}
+
+func (m *Manager) handleUpdate(update update) {
+	// Init stats
+	if _, exists := m.perSlice[update.key]; !exists {
+		m.perSlice[update.key] = &stats{}
+	}
+
+	// Update stats
+	stats := m.perSlice[update.key]
+	stats.count += 1
+	stats.size += update.size
+	if update.lastReceivedAt.After(stats.lastReceivedAt) {
+		stats.lastReceivedAt = update.lastReceivedAt
+	}
+	stats.changed = true
+}
+
+func (m *Manager) handleSync(ctx context.Context) {
+	stats := make([]model.SliceStats, 0, len(m.perSlice))
+	for k, v := range m.perSlice {
+		if v.changed {
+			stats = append(stats, model.NewSliceStats(k, v.count, v.size, v.lastReceivedAt))
+		}
+	}
+
+	m.syncFn(ctx, stats)
```

Comment:
Tu ti na konci chyba este naastavenie `stats.Changed = false`.
Idealne az po tom ako su data v etcd, kedze put moze spadnut + v goroutine ako som pisal, a teda budes potrebovat pridat zamok.

Reply:
Jo, ještě jsem se zapomněl zeptat, co mám udělat s errorem, pokud nějaký je?

Reply:
Zalogovat ako `log.Error`, pride  nam to do Slack-u a mozeme to nejak riesit.

Reply:
> Tu ti na konci chyba este nastavenie stats.Changed = false. Idealne az po tom ako su data v etcd, kedze put moze spadnut.

To nevím, jestli pomůže. Dal jsem teď ten `syncFn` spustit do vlastní goroutine. Znamená to, že i když to takhle upravím (přidat mutex nebo použít `atomic.Bool`), tak se může stát, že někdo zavolá `Notify`, nastaví se `changed = true` v `handleUpdate` a ještě než proběhne další sync, tak se dokončí save do `store` a nastaví se `changed = false`. Tím pádem se skipne jeden sync.

Pomohlo by jedině držet ten mutex do té doby, než skončí `syncFn`, ale pak nemá cenu to mít ve vlastní goroutine.

Reply:
Přidal bych tady `v.changed = false` do toho `for` loopu, kde se sbírají všechny changed stats a nechal to tak. V nejhorším případě se stane to, že nebudeme mít 100% accurate count, pokud spadne etcd. V tom případě ale asi budeme mít horší problémy.

Reply:
Ok, moze byt. Ide hlavne o to aby neposielala dookola posledna verzia, ako je to teraz.